generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java

```diff
@@ -0,0 +1,510 @@
+        operation.getTrait(HttpResponseTestsTrait.class).ifPresent(trait -> {
+            for(HttpResponseTestCase httpResponseTestCase : trait.getTestCasesFor(AppliesTo.CLIENT)){
+                generateResponseTest(operation, httpResponseTestCase);
+            }
+        });
+    }
+
+    private void generateResponseTest(OperationShape operation, HttpResponseTestCase httpResponseTestCase) {
+        if(!ProtocolTestCustomizations.TestsToSkip.contains(httpResponseTestCase.getId())){
+            if(httpResponseTestCase.getDocumentation().isPresent()){
+                writer.writeXmlDocs(httpResponseTestCase.getDocumentation().get());
+            }
+            if(ProtocolTestCustomizations.VNextTests.contains(httpResponseTestCase.getId())){
+                writer.writeSingleLineComment("This test requires a breaking change, and will be addressed in V4");
+                writer.write("[Ignore]");
+            }
+            writer.write("[TestMethod]");
+            writer.write("[TestCategory(\"ProtocolTest\")]");
+            writer.write("[TestCategory(\"ResponseTest\")]");
+            writer.write("[TestCategory(\"$L\")]",protocolNamespace);
+            writer.openBlock("public void $LResponse()\n{","}",httpResponseTestCase.getId(), () ->{
+                generateResponseTestBlock(operation,httpResponseTestCase);
+            });
+            writer.write("\n");
+        }
+    }
+
+    private void generateResponseTestBlock(OperationShape operation, HttpResponseTestCase httpResponseTestCase) {
+        var outputShape =  model.expectShape(operation.getOutputShape(), StructureShape.class);
+        var responseSymbol = operation.getId().getName() + "Response";
+        //Arrange
+        arrangeResponseTestBlock(httpResponseTestCase);
+        writer.write("\n");
+        // only unmarshall the response and assert if a body is present, as per smithy spec
+        if(httpResponseTestCase.getBody().isPresent() && !httpResponseTestCase.getBody().equals("")){
+            //Act
+            writer.writeSingleLineComment("Act");
+            writer.write("var unmarshalledResponse = new $LUnmarshaller().Unmarshall(context);",responseSymbol);
+            writer.openBlock("var expectedResponse = new $L\n{","};",responseSymbol,(Runnable) () -> httpResponseTestCase.getParams().accept(new ValueNodeVisitor(outputShape,true, responseSymbol)));
+            writer.write("\n");
+            //Assert
+            writer.writeSingleLineComment("Assert");
+            writer.write("var actualResponse = ($L)unmarshalledResponse;",responseSymbol);
+            writer.write("Comparer.CompareObjects<$L>(expectedResponse,actualResponse);", responseSymbol);
+        }
+        writer.write("Assert.AreEqual((HttpStatusCode)Enum.ToObject(typeof(HttpStatusCode), $L), context.ResponseData.StatusCode);",httpResponseTestCase.getCode());
+    }
+
+    private final class ValueNodeVisitor implements NodeVisitor<Void> {
```

Comment:
i didn't pay super attention to this as it depends on how types are converted in C# which I'm not familiar with